amethyst_ui/src/event_retrigger.rs

```diff
@@ -0,0 +1,64 @@
+use std::default::Default;
+
+use amethyst_core::{
+    shrev::{EventChannel, ReaderId},
+    specs::prelude::{
+        Component, Read, Write, ReadStorage, Resources, System, SystemData,
+    },
+};
+
+use event::TargetedEvent;
+
+pub trait EventRetrigger: Component {
+    type In: Clone + Send + Sync + TargetedEvent;
+    type Out: Send + Sync;
+
+    fn apply(&self, event: &Self::In) -> Vec<Self::Out>;
```

Comment:
This signature requires the implementation to allocate a vector. An alternative could be to use a different trait to describe how data can be read out. That way the receiver can decide:

```rust
pub trait Receiver<T> {
    fn receive(&mut self, values: &[T]);
}

impl<T> Receiver<T> for Vec<T>
where
    T: Clone,
{
    fn receive(&mut self, values: &[T]) {
        self.extend(values.iter().cloned());
    }
}

pub trait Trigger {
    type In;
    type Out: Send + Sync;

    fn apply<R>(&self, event: &Self::In, out: &mut R)
    where
        R: Receiver<Self::Out>;
}

pub struct Example;

impl Trigger for Example {
    type In = ();
    type Out = u32;

    fn apply<R>(&self, event: &Self::In, out: &mut R)
    where
        R: Receiver<Self::Out>,
    {
        out.receive(&[42]);
    }
}

fn main() {
    let mut out = Vec::new();
    Example.apply(&(), &mut out);
    println!("out: {:?}", out);
}
```

Reply:
That's actually a very good idea! I was struggling a lot with how to return these events, since returning a `Vec` always feels a little bad, especially in this context. I'll try this out, but maybe it might also be a good idea to somehow pass the `EventChannel` to it, so that it can write it's items there directly? That would reduce the overhead of having to construct some sort of container on the receiving end

Reply:
Done!